### PR TITLE
Add documentation for NLU training

### LINKED ISSUE #241: Add documentation for NLU training
Create a page describing how to train JoinBERT.
A link to this page should be added to `README.md`.

### PR DESCRIPTION
**Pull request by IKostric**
Fixes #241

### CODE REVIEW

File: docs/source/architecture.rst [hunk 1]
```diff
@@ -11,7 +11,41 @@ The system architecture is shown in the figure below, illustrating the core proc
 Natural Language Understanding
 ------------------------------
 
-The :py:class:`NLU <moviebot.nlu.nlu>` component converts the natural language :py:class:`UserUtterance <moviebot.core.utterance.utterance.UserUtterance>` into a :py:class:`DialogueAct <moviebot.dialogue_manager.dialogue_act>`. This process, comprising of *intent detection* and *slot filling*, is performed based on the current dialogue state.
+The :py:class:`NLU <moviebot.nlu.nlu>` component converts the natural language :py:class:`UserUtterance <moviebot.core.utterance.utterance.UserUtterance>` into a :py:class:`DialogueAct <moviebot.dialogue_manager.dialogue_act>`. This process, comprising of *intent detection* and *slot filling*, is performed based on the current dialogue state. The component offers two distinct modules: Rule-Based and Neural using JointBERT.
```
Comment (kbalog) [inline docs/source/architecture.rst:14]: Nit: "two distinct modules" => "two distinct solutions as modules"

File: docs/source/architecture.rst [hunk 2]
```diff
@@ -11,7 +11,41 @@ The system architecture is shown in the figure below, illustrating the core proc
...
+
+Rule-Based NLU
+^^^^^^^^^^^^^^
+
+The rule-based NLU module utilizes a combination of keyword extraction and heuristics to determine the user's intent and extract slot-value pairs from their utterances. This approach relies on predefined rules and patterns to interpret user input.
+
+Neural NLU with JointBERT
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The Neural NLU module employs JointBERT, a neural model trained specifically for intent detection and slot filling tasks. JointBERT predicts both the intent of the user's utterance and the corresponding slot-value pairs using a trained model.
```
Comment (kbalog) [inline docs/source/architecture.rst:24]: The second sentence is redundant. Can be simplified:
"The Neural NLU module employs JointBERT, a neural model trained for predicting both the intent of the user's utterance and the corresponding slot-value pairs."

File: docs/source/architecture.rst [hunk 3]
```diff
@@ -11,7 +11,41 @@ The system architecture is shown in the figure below, illustrating the core proc
...
+
+Training the JointBERT Model
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To train the JointBERT model, the provided training script (`/Users/2920807/Repos/moviebot/moviebot/nlu/annotation/joint_bert/joint_bert_train.py`) can be utilized. This script fine-tunes the pre-trained BERT model on a dataset annotated with intents and slot-value pairs. Below is an overview of the training process:
```
Comment (kbalog) [inline docs/source/architecture.rst:29]: Fix local path to relative path within repo

File: docs/source/architecture.rst [hunk 4]
```diff
@@ -11,7 +11,41 @@ The system architecture is shown in the figure below, illustrating the core proc
...
+Example dataset format:
+```yaml
+REVEAL:
+  - text: "[I absolutely adore](modifier) movies focusing on [martial arts](keywords)."
+  - text: "Films about [space exploration](keywords) [fascinate me](modifier)."
+  - text: "[I can't stand](modifier) movies that emphasize on [corporate politics](keywords)."
+  - text: "[Space adventures](keywords) [always intrigue me](modifier)."
+```
+
+2. **Model Initialization**: The model is initialized with the number of intent labels and slot labels based on the dataset. Additionally, you can configure hyperparameters such as learning rate, weight decay, and maximum epochs.
```
Comment (kbalog) [inline docs/source/architecture.rst:42]: Nit: "Additionally, you can configure hyperparameters ..." => "Additionally, hyperparameters ... may be configured."

File: docs/source/architecture.rst [hunk 5]
```diff
@@ -11,7 +11,41 @@ The system architecture is shown in the figure below, illustrating the core proc
...
+
+3. **Training**: The training script supports logging with Wandb for easy monitoring of training progress.
```
Comment (kbalog) [inline docs/source/architecture.rst:44]: Add link to Wandb

File: docs/source/architecture.rst (hunk 3 above)
Comment (kbalog) [inline docs/source/architecture.rst:29]: Does the script download the base BERT model or is that needed as "step 0"?

**Review by kbalog — CHANGES_REQUESTED** [review]

File: docs/source/architecture.rst (hunk 1 above)
Comment (IKostric) [reply docs/source/architecture.rst:14]: Done

File: docs/source/architecture.rst (hunk 2 above)
Comment (IKostric) [reply docs/source/architecture.rst:24]: Done

File: docs/source/architecture.rst (hunk 3 above)
Comment (IKostric) [reply docs/source/architecture.rst:29]: Done

Comment (IKostric) [reply docs/source/architecture.rst:29]: It downloads the pre-trained BERT model automatically.

File: docs/source/architecture.rst (hunk 4 above)
Comment (IKostric) [reply docs/source/architecture.rst:42]: Done

File: docs/source/architecture.rst (hunk 3 above)
Comment (kbalog) [reply docs/source/architecture.rst]: Then just mention it: "This script downloads a pre-trained BERT model and fine-tunes it on a dataset..."

**Review by kbalog — APPROVED** [review]
LGTM

**Pull request merged**